awscli/botocore/utils.py

```diff
@@ -3305,6 +3339,231 @@ def fetch_token(
+
+        return registration
+
+    def _registration(
+            self,
+            start_url,
+            session_name,
+            scopes,
+            force_refresh=False,
+    ):
+        cache_key = self._registration_cache_key(
+            start_url,
+            session_name,
+            scopes,
+        )
+        if not force_refresh and cache_key in self._cache:
+            registration = self._cache[cache_key]
+            if (not self._is_expired(registration) and
+                    self._is_registration_for_auth_code(registration)):
+                return registration
+
+        registration = self._register_client(
+            session_name,
+            scopes,
+            self._auth_code_fetcher.redirect_uri_without_port(),
+            start_url
+        )
+        self._cache[cache_key] = registration
+        return registration
+
+    def _extract_resolved_endpoint(self, params, **kwargs):
+        """Event handler for before-call that will extract the resolved endpoint
+        for a given request without actually running it
+        """
+        # This will contain any path and query params specific to
+        # the operation/input, so extract just the scheme and hostname
+        if params['url']:
+            parsed = urlparse(params['url'])
+            self._base_endpoint = f'{parsed.scheme}://{parsed.netloc}'
+
+        # Return a tuple containing the "response" to short-circuit the request
+        return botocore.awsrequest.AWSResponse(None, 200, {}, None), {}
+
+    def _get_base_authorization_uri(self):
+        """Simulates an SSO-OIDC request so that we can extract the "base"
+        endpoint for the current client to use for the un-modeled Authorize
+        operation
+        """
+        self._client.meta.events.register('before-call',
+                                          self._extract_resolved_endpoint)
+        self._client.register_client(
+            clientName='temp',
+            clientType='public'
+        )
+        self._client.meta.events.unregister('before-call',
+                                            self._extract_resolved_endpoint)
+
+        return self._base_endpoint
+
+    def _get_authorization_uri(
+            self,
+            client_id,
+            registration_scopes,
+            expected_state):
+
+        query_params = {
+            'response_type': 'code',
+            'client_id': client_id,
+            'redirect_uri': self._auth_code_fetcher.redirect_uri_with_port(),
+            'state': expected_state,
+            'code_challenge_method': 'S256'
+            # Don't want to encode code_challenge again, so we append below
+        }
+
+        # For the query param, scopes must be space separated before encoding
+        if registration_scopes:
+            query_params['scope'] = " ".join(registration_scopes)
+
+        return (
+            f'{self._get_base_authorization_uri()}/authorize?'
+            f'{urllib.parse.urlencode(query_params)}'
```

Comment:
Let's import `urlencode` from `botocore.compat` to keep it referencing one place.

Reply:
Done via https://github.com/aws/aws-cli/pull/8947/commits/e306e3ec67d3f8e3fb1d419f8d148f13f53f16a4